README
PathConvert tools

						2016/11/24 Shigio Yamaguchi

This is a package which includes tools for path conversion.

	Contents
	========
	1. Pathconvert library for C language
	2. Improved X lndir
	3. Improved BSD ln(1)
	4. Improved GNU ln(1)
	5. File-PathConvert module for perl5


1. Pathconvert library for C language
=====================================

Two functions are available.

o abs2rel: make a relative path name from an absolute path name

  abs2rel(<original path name>, <base directory>, <result>, <size of result>);

	INPUT			INPUT			OUTPUT
        <original path name>    <base directory>        <result>
        /usr/src                /etc                    ../usr/src


o rel2abs: make an absolute path name from a relative path name

  rel2abs(<original path name>, <base directory>, <result>, <size of result>);

	INPUT			INPUT			OUTPUT
        <original path name>    <base directory>        <result>
        ../usr/src              /etc                    /usr/src

How to install:

	% cd lib				# move to lib directory
	% make					# make
	% make test				# test


2. Improved X lndir
===================

Lndir duplicates directory tree and makes symbolic links to original files.
This command is derived from X11 and slightly modified to generate relative
symbolic link.

Installation:

        % make
	% cd lndir
        % xmkmf
        % make
        % make install

Usage:

        % mkdir /tmp/sys_shadow
        % lndir /usr/src/sys /tmp/sys_shadow
        /usr/src/sys/compile:
        /usr/src/sys/conf:
        /usr/src/sys/ddb:
        /usr/src/sys/dev:
                .
                .
                .
        % cd /tmp/sys_shadow
        % ls -lL Makefile
        -rw-r--r--  1 root  wheel  190 Jun 14  1996 Makefile
        % ls -l Makefile
	lrwxr-xr-x  1 shigio  wheel  26 Dec 29 14:18 Makefile -> ../../usr/src/sys/Makefile

Using original lndir, it would be absolute like this.

        lrwxr-xr-x  1 shigio  wheel  26 Dec 29 14:18 Makefile -> /usr/src/sys/Makefile


3. Improved BSD ln(1)
=====================

This command is derived from FreeBSD and slightly modified to make right and
relative (or absolute) symbolic link. -a and -r option are added.

Installation:

	(You need BSD build system.)
        % make
	% cd ln
        % make
        % make install

Ex1: You can make right and relative (or absolute) symbolic link.
	% cd /usr/src/sys
	% ls -l kern/tty.c
	-rw-r--r--  1 root  wheel  56525 Jul  3 09:27 kern/tty.c
	% ln -sf kern/tty.c ~/tmp
	% ls -l ~/tmp/tty.c
	lrwxrwxrwx  1 shigio  users  10 Jul 13 18:08 /home/shigio/tmp/tty.c ->
					kern/tty.c
					~~~~~~~~~~
					(wrong symbolic link)
	% ln -sfr kern/tty.c ~/tmp
	% ls -l ~/tmp/tty.c
	lrwxrwxrwx  1 shigio  users  31 Jul 13 13:28 /home/shigio/tmp/tty.c ->
					../../../usr/src/sys/kern/tty.c
					~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
					(right and relative symbolic link)
	% ln -sfa kern/tty.c ~/tmp
	% ls -l ~/tmp/tty.c
	lrwxrwxrwx  1 shigio  users  31 Jul 13 13:28 /home/shigio/tmp/tty.c ->
					/usr/src/sys/kern/tty.c
					~~~~~~~~~~~~~~~~~~~~~~~
					(right and absolute symbolic link)

Ex2: You can make symbolic link even if the target isn't exist.
     (This is a spec of symbolic link. ln(1) doens't print warning message.)
	% cd /usr/src/sys
	% ls -l kern/noexist.c
	ls: kern/noexist.c: No such file or directory
	% ln -sr kern/noexist.c ~/tmp
	% ls -l ~/tmp/noexist.c 
	lrwxrwxrwx  1 shigio  users  36 Jul 13 17:53 /home/shigio/tmp/noexist.c ->
					../../../usr/src/sys/kern/noexist.c
					~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
					(not exist but right symbolic link)

Ex3: You can make shadow tree with relative symbolic link like this.
	% cd /usr/src/sys
	% set obj=/usr/obj`pwd`
	% mkdir -p $obj
	% find * -type d -exec mkdir $obj/'{}' ';'
	% find * -type f -exec ln -sr '{}' $obj/'{}' ';'


4. Improved GNU ln(1)
=====================

This command is derived from GNU fileutils-3.16 and slightly modified to make
right and relative (or absolute) symbolic link. -a and -r option are added.

Installation:

        (Please fetch fileutils-3.16 from ftp site. Then ...)
	% cd gnuln
	% tar xzf fileutils-3.16.tar.gz
	% cp ../lib/abs2rel.c ../lib/rel2abs.c fileutils-3.16/lib
	% cp Makefile.in fileutils-3.16/lib
	% cp ln.c fileutils-3.16/src
	% cp ln.1 fileutils-3.16/man
        % cd fileutils-3.16
        % sh configure
	% make
	# make install

Usage is the same with improved BSD ln(1).


5. File-PathConvert module for perl5
====================================

++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
NOTE:	Now File::PathConvert module was merged into File::Spec module
	which is included by perl core package.  So, you had better
	use File::Spec instead of this module.
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

This is a module to use path conversion facility from perl5 environment.
You need perl version 5.002 or later.

This package includes two version of File-PathConvert.

	o File-PathConvert-0.4		unix version
	o File-PathConvert-0.85		multiplatform version

Installation:

        % perl Makefile.PL
        % make
        % make test
        # make install          <- as a root

This module includes sample application 'mkshadow'.
You will find the usage for the looking.

----------------------------------------------------------------------------
The latest verion is here.

	http://www.tamacom.com/pathconvert.html

Comments appreciated. Thank you in advance.


ChangeLog
Pathconvert project history.
----------------------------

23-Nov-2016	Pathconvert library was adopted by FreeBSD core (lib/libpathconv).

11-Jul-2006	Removed abs2rel (bash script) because it is not maintained well.

19-Dec-2005	Added LGPL version of abs2rel() and rel2abs().

29-Aug-2004	Replaced Makefile with more generic one and fixed README. 

9-Feb-2004	Bug fixed: rel2abs(".", "/") generated ''. It should be '/'.
		Reported by Sergei Birj.

18-May-2002	abs2rel script by D. J. Hawkey Jr.

27-April-2002	RFC about g_abs2rel and g_rel2abs on 'gtk-devel-list gnome org'
		mailing list.

		From: "Shigio Yamaguchi" <shigio tamacom com> 
		To: gtk-devel-list gnome org 
		Subject: New functions for converting path name. 
		Date: Sat, 27 Apr 2002 12:58:23 +0900 

14-Feb-2002	Abs2rel() function was merged into GNU GLOBAL source code
		tag system.

3-Sep-1999	NetBSD includes File::PathConvert version 0.85 as a package.

15-Feb-1999	File::PathConvert module was merged into File::Spec as a part
		of perl5.005_55 (development release working toward 5.006). 

3-Feb-1999	File::PathConvert version 0.85 by Barrie Slaymaker.

18-Dec-1998	File::PathConvert version 0.84 by Barrie Slaymaker.

4-Dec-1998	Improved GNU ln(1).

21-Nov-1998	File::PathConvert version 0.82 by Barrie Slaymaker.

		Added beginnings of URL support.

8-Sep-1998	File::PathConvert version 0.7 by Barrie Slaymaker.

		Implemented a more flexible parameterization scheme for 
		customizing for different OSs.  WE NEED MacOS, VMS, OS/2, etc.

5-Sep-1998	File::PathConvert version 0.6 by Barrie Slaymaker.

		Incorporated feedback from Tye McQueen to make REs more
		reliable and clear.

30-Aug-1998	File::PathConvert version 0.5 by Barrie Slaymaker.

		Added support for multiplatform use, fully implemented and
		tested only for Win32 based perl.

		1st RFC of multiplatform version.

		Newsgroups: comp.lang.perl,comp.lang.perl.modules
		Message-ID: <35E985CD.AD289186@lm.com>
		Date: Sun, 30 Aug 1998 13:03:09 -0400
		From: "Barrie Slaymaker (home)" <rbs@lm.com>
		Subject: File::PathConvert for your review

14-Jul-1998	Improved BSD ln(1) 2nd RFC.

		To: hackers@FreeBSD.ORG
		Subject: Improvement of ln(1). [Second RFC]
		Date: Tue, 14 Jul 1998 00:40:17 +0900

13-Jul-1998	Improved BSD ln(1).

11-Jul-1998	Improved BSD ln(1) 1st RFC.

		To: hackers@freebsd.org
		Subject: Improvemnet of ln(1).
		Date: Sat, 11 Jul 1998 02:39:37 +0900

8-Mar-1998	File::PathConvert for perl5 version 0.4

		mkshadow
			- Bug (Mkshadow converted symbolic links into real path.)
			  fixed.
		abs2rel	- modified to produce regularize path name.


9-Jan-1998	File::PathConvert for perl5 version 0.3

		PathConvert.pm
			- fixed the statment of online manual.

		Following useful modifications were submitted by Jesse N. Glick.

		* using $VERSION, to placate CPAN, Exporter, MakeMaker, etc.
		* exporting package variables, all of which seem to be designed
		  for possible customization (or retrieval of info in the case
		  of $resolved)
		* prototyping public functions (those extra semicolons prevent
		  e.g. CPerl mode from thinking `$' escapes the following `)'
		  and messing up Emacs)
		* using standard indented format for literal program examples;
		  use of .br is nonstandard and will not work on anything but
		  nroff/troff (e.g. pod2html)
		* including suggested use of $resolved in SYNOPSIS
		* putting more useful information in the warning messages

29-Dec-1997	Pathconvert library for C language alpha 3 version.

		Sample application lndir was added.

22-Dec-1997	Pathconvert library for C language 2nd RFC
		and alpha 2nd version.

		To: hackers@freebsd.org
		Subject: second RFC about abs2rel() and rel2abs().
		Date: Mon, 29 Dec 1997 15:10:14 +0900

		SYNOPSIS was changed like this.

		#include <sys/param.h>
		char *
		abs2rel(const char *path, const char *base,
						char result[MAXPATHLEN])
			|
			v
		char *
		abs2rel(const char *path, const char *base,
						char *result, size_t size)
						~~~~~~~~~~~~  ~~~~~~~~~~~
						buffer address, buffer size

		#include <sys/param.h>
		char *
		rel2abs(const char *path, const char *base,
						char result[MAXPATHLEN])
			|
			v
		char *
		rel2abs(const char *path, const char *base,
						char *result, size_t size)
						~~~~~~~~~~~~  ~~~~~~~~~~~
						buffer address, buffer size

		Thanks to
			Andrew Kenneth Milton
			John Polstra
		for their comments.

4-Dec-1997	Pathconvert library for C language 1st RFC
		and alpha 1st version.

		Fist RFC on hackers@freebsd.org.
		To: hackers@freebsd.org
		Message-Id: <199712032230.WAA28837@wafu.netgate.net>
		Subject: [RFC] path converting functions.
		Date: Thu, 04 Dec 1997 15:28:35 +0900

3-Sep-1997	File::PathConvert for perl5 version 0.2

		PathConvert.pm
			- SYNOPSIS fixed.
			- Tested with perl5.004_02.
		README	- Mail address changed.
			- History added.

23-Oct-1996	File::PathConvert for perl5 version 0.1

			Tested with perl5.003_02.

7-Oct-1996	File::PathConvert for perl5 2nd RFC.

		second RFC on comp.lang.perl.modules.
		Subject: 2nd RFC File::PathConvert module
		Message-ID: <53a449$2qo@newsgate.so-net.or.jp>

		PathConvert.pm -
			use @EXPORT_OK instead of @EXPORT.
			realpath added.
			use $SL variable for path separator.

17-Sep-1996	File::PathConvert for perl5 1st RFC.

		first RFC on comp.lang.perl.modules.
		Subject: RFC File::PathConvert module
		Message-ID: <51m90o$pos@newsgate.so-net.or.jp>


lib/t/test.pl
#!/usr/bin/perl
#
# Test script for abs2rel(3) and rel2abs(3).
#
$logfile = 'err';
#
#       target          base directory  result
#       --------------------------------------
@abs2rel = (
	'.		/		.',
	'a/b/c		/		a/b/c',
	'a/b/c		/a		a/b/c',
	'/a/b/c		a		ERROR',
);
@rel2abs = (
	'.		/a		/a',
	'.		/		/',
	'./		/		/',
	'/a/b/c		/		/a/b/c',
	'/a/b/c		/a		/a/b/c',
	'a/b/c		a		ERROR',
	'..		/a		/',
	'../		/a		/',
	'../..		/a		/',
	'../../		/a		/',
	'../../..	/a		/',
	'../../../	/a		/',
	'../b		/a		/b',
	'../b/		/a		/b/',
	'../../b	/a		/b',
	'../../b/	/a		/b/',
	'../../../b	/a		/b',
	'../../../b/	/a		/b/',
	'../b/c		/a		/b/c',
	'../b/c/	/a		/b/c/',
	'../../b/c	/a		/b/c',
	'../../b/c/	/a		/b/c/',
	'../../../b/c	/a		/b/c',
	'../../../b/c/	/a		/b/c/',
);
@common = (
	'/a/b/c		/a/b/c		.',
	'/a/b/c		/a/b/		c',
	'/a/b/c		/a/b		c',
	'/a/b/c		/a/		b/c',
	'/a/b/c		/a		b/c',
	'/a/b/c		/		a/b/c',
	'/a/b/c		/a/b/c		.',
	'/a/b/c		/a/b/c/		.',
	'/a/b/c/	/a/b/c		./',
	'/a/b/		/a/b/c		../',
	'/a/b		/a/b/c		..',
	'/a/		/a/b/c		../../',
	'/a		/a/b/c		../..',
	'/		/a/b/c		../../../',
	'/a/b/c		/a/b/z		../c',
	'/a/b/c		/a/y/z		../../b/c',
	'/a/b/c		/x/y/z		../../../a/b/c',
);
print "TEST start ";
open(LOG, ">$logfile") || die("cannot open log file '$logfile'.\n");
$cnt = 0;
$progname = 'abs2rel';
foreach (@abs2rel) {
	@d = split;
	chop($result = `./$progname $d[0] $d[1]`);
	if ($d[2] eq $result) {
		print '.';
	} else {
		print 'X';
		print LOG "$progname $d[0] $d[1] -> $result (It should be '$d[2]')\n";
		$cnt++;
	}
}
foreach (@common) {
	@d = split;
	chop($result = `./$progname $d[0] $d[1]`);
	if ($d[2] eq $result) {
		print '.';
	} else {
		print 'X';
		print LOG "$progname $d[0] $d[1] -> $result (It should be '$d[2]')\n";
		$cnt++;
	}
}
$progname = 'rel2abs';
foreach (@rel2abs) {
	@d = split;
	chop($result = `./$progname $d[0] $d[1]`);
	if ($d[2] eq $result) {
		print '.';
	} else {
		print 'X';
		print LOG "$progname $d[0] $d[1] -> $result (It should be '$d[2]')\n";
		$cnt++;
	}
}
foreach (@common) {
	@d = split;
	chop($result = `./$progname $d[2] $d[1]`);
	if ($d[0] eq $result) {
		print '.';
	} else {
		print 'X';
		print LOG "$progname $d[2] $d[1] -> $result (It should be '$d[0]')\n";
		$cnt++;
	}
}
close(LOG);
if ($cnt == 0) {
	print " COMPLETED.\n";
} else {
	print " $cnt errors detected.\n";
	open(LOG, $logfile) || die("log file not found.\n");
	while (<LOG>) {
		print;
	}
	close(LOG);
}


lib/t/abs2rel.c
#include <stdio.h>
#include <sys/param.h>
#include <errno.h>

char	*abs2rel __P((const char *, const char *, char *, size_t));

main(argc, argv)
int	argc;
char	*argv[];
{
	char result[MAXPATHLEN];
	char cwd[MAXPATHLEN];

	if (argc < 2) {
		fprintf(stderr, "usage: abs2rel path [base]\n");
		exit(1);
	}
	if (argc == 2) {
		if (!getcwd(cwd, MAXPATHLEN)) {
			fprintf(stderr, "cannot get current directory.\n");
			exit(1);
		}
	} else
		strcpy(cwd, argv[2]);
	
	if (abs2rel(argv[1], cwd, result, MAXPATHLEN)) {
		printf("%s\n", result);
	} else
		printf("ERROR\n");
	exit(0);
}


lib/t/rel2abs.c
#include <stdio.h>
#include <sys/param.h>
#include <errno.h>

char	*rel2abs __P((const char *, const char *, char *, size_t));

main(argc, argv)
int	argc;
char	*argv[];
{
	char result[MAXPATHLEN];
	char cwd[MAXPATHLEN];

	if (argc < 2) {
		fprintf(stderr, "usage: rel2abs path [base]\n");
		exit(1);
	}
	if (argc == 2) {
		if (!getcwd(cwd, MAXPATHLEN)) {
			fprintf(stderr, "cannot get current directory.\n");
			exit(1);
		}
	} else
		strcpy(cwd, argv[2]);
	
	if (rel2abs(argv[1], cwd, result, MAXPATHLEN)) {
		printf("%s\n", result);
	} else
		printf("ERROR\n");
	exit(0);
}


lib/abs2rel.c
/*
 * Copyright (c) 1997 Shigio Yamaguchi. All rights reserved.
 * Copyright (c) 1999 Tama Communications Corporation. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */
#include <errno.h>
#include <stdlib.h>
#include <string.h>
/*
 * abs2rel: convert an absolute path name into relative.
 *
 *	i)	path	absolute path
 *	i)	base	base directory (must be absolute path)
 *	o)	result	result buffer
 *	i)	size	size of result buffer
 *	r)		!= NULL: relative path
 *			== NULL: error
 */
char *
abs2rel(path, base, result, size)
	const char *path;
	const char *base;
	char *result;
	const size_t size;
{
	const char *pp, *bp, *branch;
	/*
	 * endp points the last position which is safe in the result buffer.
	 */
	const char *endp = result + size - 1;
	char *rp;

	if (*path != '/') {
		if (strlen(path) >= size)
			goto erange;
		strcpy(result, path);
		goto finish;
	} else if (*base != '/' || !size) {
		errno = EINVAL;
		return (NULL);
	} else if (size == 1)
		goto erange;
	/*
	 * seek to branched point.
	 */
	branch = path;
	for (pp = path, bp = base; *pp && *bp && *pp == *bp; pp++, bp++)
		if (*pp == '/')
			branch = pp;
	if ((*pp == 0 || (*pp == '/' && *(pp + 1) == 0)) &&
	    (*bp == 0 || (*bp == '/' && *(bp + 1) == 0))) {
		rp = result;
		*rp++ = '.';
		if (*pp == '/' || *(pp - 1) == '/')
			*rp++ = '/';
		if (rp > endp)
			goto erange;
		*rp = 0;
		goto finish;
	}
	if ((*pp == 0 && *bp == '/') || (*pp == '/' && *bp == 0))
		branch = pp;
	/*
	 * up to root.
	 */
	rp = result;
	for (bp = base + (branch - path); *bp; bp++)
		if (*bp == '/' && *(bp + 1) != 0) {
			if (rp + 3 > endp)
				goto erange;
			*rp++ = '.';
			*rp++ = '.';
			*rp++ = '/';
		}
	if (rp > endp)
		goto erange;
	*rp = 0;
	/*
	 * down to leaf.
	 */
	if (*branch) {
		if (rp + strlen(branch + 1) > endp)
			goto erange;
		strcpy(rp, branch + 1);
	} else
		*--rp = 0;
finish:
	return result;
erange:
	errno = ERANGE;
	return (NULL);
}


p5module/File-PathConvert-0.4/PathConvert.pm
#
# Copyright (c) 1996, 1997, 1998 Shigio Yamaguchi. All rights reserved.
#
# This program is free software; you can redistribute it and/or
# modify it under the same terms as Perl itself.
#
#	File::PathConvert.pm
#
#				8-Mar-1998 Shigio Yamaguchi
#
package File::PathConvert;
$VERSION = '0.4';

require 5.002;
require Exporter;
@ISA = qw(Exporter);
@EXPORT_OK = qw(realpath abs2rel rel2abs $maxsymlinks $verbose $SL $resolved);
use Cwd;
#
# instant configration
#
$maxsymlinks = 32;		# allowed symlink number in a path
$verbose = 0;			# 1: verbose on, 0: verbose off
$SL = '/';			# separator
#
# realpath: returns the canonicalized absolute path name
#
# Interface:
#	i)	$path	path
#	r)		resolved name on success else undef
#	go)	$resolved
#			resolved name on success else the path name which
#			caused the problem.
	$resolved = '';
#
#	Note: this implementation is based 4.4BSD version realpath(3).
#
sub realpath($;) {
    ($resolved) = @_;
    my($backdir) = cwd();
    my($dirname, $basename, $links, $reg);

    regularize($resolved);
LOOP:
    {
	#
	# Find the dirname and basename.
	# Change directory to the dirname component.
	#
	if ($resolved =~ /$SL/) {
	    $reg = '^(.*)' . $SL . '([^' . $SL . ']*)$';
	    ($dirname, $basename) = $resolved =~ /$reg/;
	    $dirname = $SL if (!$dirname);
	    $resolved = $dirname;
	    unless (chdir($dirname)) {
		warn("realpath: chdir($dirname) failed: $! (in ${\cwd()}).") if $verbose;
		chdir($backdir);
		return undef;
	    }
	} else {
	    $dirname = '';
	    $basename = $resolved;
	}
	#
	# If it is a symlink, read in the value and loop.
	# If it is a directory, then change to that directory.
	#
	if ($basename) {
	    if (-l $basename) {
		unless ($resolved = readlink($basename)) {
		    warn("realpath: readlink($basename) failed: $! (in ${\cwd()}).") if $verbose;
		    chdir($backdir);
		    return undef;
		}
		$basename = '';
		if (++$links > $maxsymlinks) {
		    warn("realpath: too many symbolic links: $links.") if $verbose;
		    chdir($backdir);
		    return undef;
		}
		redo LOOP;
	    } elsif (-d _) {
		unless (chdir($basename)) {
		    warn("realpath: chdir($basename) failed: $! (in ${\cwd()}).") if $verbose;
		    chdir($backdir);
		    return undef;
		}
		$basename = '';
	    }
	}
    }
    #
    # Get the current directory name and append the basename.
    #
    $resolved = cwd();
    if ($basename) {
	$resolved .= $SL if ($resolved ne $SL);
	$resolved .= $basename
    }
    chdir($backdir);
    return $resolved;
}
#
# abs2rel: make a relative pathname from an absolute pathname
#
# Interface:
#	i)	$path	absolute path(needed)
#	i)	$base	base directory(optional)
#	r)		relative path of $path
#
#	Note:	abs2rel doesn't check whether the specified path exist or not.
#
sub abs2rel($;$;) {
    my($path, $base) = @_;
    my($reg, $common);

    $reg = '^' . $SL;
    if ($path !~ /$reg/) {
	warn("abs2rel: nothing to do.($path)") if $verbose;
	return $path;
    }
    if (!$base) {
	$base = cwd();
    } elsif ($base !~ /$reg/) {
	$base = cwd() . $SL . $base;
    }
    regularize($path);
    regularize($base);
    $reg = $SL . '[^' . $SL . ']+' . $SL . '\.\.';
    while ($base =~ /$reg/) {
	$base =~ s/$reg//;			# trim path/..
    }
    $common = common($path, $base);
    $path =~ s/$common//;
    $base =~ s/$common//;
    $reg = '^'. $SL;
    $path =~ s/$reg//;
    $base =~ s/$reg//;
    $reg = '[^' . $SL . ']+';
    $base =~ s/$reg/../g;
    if (!($path . $base)) {
	$path = '.';
    } elsif ($path && $base) {
	$path = $base . $SL . $path;
    } else {
	$path = $base . $path;
    }
    regularize($path);
    $path;
}

#
# rel2abs: make an absolute pathname from a relative pathname
#
# Interface:
#	i)	$path	relative path (needed)
#	i)	$base	base directory 	(optional)
#	r)		absolute path of $path
#
#	Note:	rel2abs doesn't check whether the specified path exist or not.
#
sub rel2abs($;$;) {
    my($path, $base) = @_;
    my($reg);

    $reg = '^' . $SL;
    if ($path =~ /$reg/) {
	warn("rel2abs: nothing to do.($path)") if $verbose;
        return $path;
    }
    regularize($path);
    if (!$base) {
        $base = cwd();
    }
    $reg = '^' . $SL;
    if ($base !~ /$reg/) {
	$base = cwd() . '/' . $base;
    }
    $path = $base . $SL . $path;
    $reg = $SL . '[^' . $SL . ']+' . $SL . '\.\.';
    while ($path =~ /$reg/) {
	$path =~ s/$reg//;			# trim path/..
    }
    regularize($path);
    $path;
}

#
# regularize a path.
#
sub regularize {
    my($reg);

    $reg = '^' . $SL . '\.\.' . $SL;
    while ($_[0] =~ /$reg/) {		# ^/../	-> /
	$_[0] =~ s/$reg/$SL/;
    }
    $reg = $SL . '\.' . $SL;
    while ($_[0] =~ /$reg/) {
	$_[0] =~ s/$reg/$SL/;		# /./ -> /
    }
    $reg = $SL . '+';
    $_[0] =~ s/$reg/$SL/g;		# ///  -> /
    $reg = '(.+)' . $SL . '$';
    $_[0] =~ s/$reg/$1/;		# remove last /
    $reg = '(.+)' . $SL . '\.$';
    $_[0] =~ s/$reg/$1/g;		# remove last /.
    $_[0] = '/' if $_[0] eq '/.';
}

#
# extract common part of two paths.
#
sub common {
    my($p1, $p2) = @_;
    my(@p1, @p2, @common);

    @p1 = split($SL, $p1);
    @p2 = split($SL, $p2);
    while (@p1 && @p2 && $p1[0] eq $p2[0]) {
	push @common, shift @p1;
	shift @p2;
    }
    join($SL, @common);
}

1;

__END__

=head1 NAME

realpath - make a canonicalized absolute path name

abs2rel - make a relative path from an absolute path

rel2abs - make an absolute path from a relative path

=head1 SYNOPSIS

    use File::PathConvert qw(realpath abs2rel rel2abs);

    $path = realpath($path);

    $path = abs2rel($path);
    $path = abs2rel($path, $base);

    $path = rel2abs($path);
    $path = rel2abs($path, $base);

    use File::PathConvert qw($resolved);
    $path = realpath($path) || die "resolution stopped at $resolved";

=head1 DESCRIPTION

The PathConvert module provides three functions.

=over 4

=item realpath

C<realpath> makes a canonicalized absolute pathname and
resolves all symbolic links, extra ``/'' characters, and references
to /./ and /../ in the path.
C<realpath> resolves both absolute and relative paths.
It returns the resolved name on success, otherwise it returns undef
and sets the valiable C<$File::PathConvert::resolved> to the pathname
that caused the problem.

All but the last component of the path must exist.

This implementation is based on 4.4BSD realpath(3).

=item abs2rel

C<abs2rel> makes a relative path name from an absolute path name.
By default, the base is the current directory.
If you specify a second parameter, it's assumed to be the base.

The returned path may include symbolic links.
C<abs2rel> doesn't check whether or not any path exists.

=item rel2abs

C<rel2abs> makes an absolute path name from a relative path name.
By default, the base directory is the current directory.
If you specify a second parameter, it's assumed to be the base.

The returned path may include symbolic links.
C<abs2rel> doesn't check whether or not any path exists.

=head1 EXAMPLES

=item realpath

    If '/sys' is a symbolic link to '/usr/src/sys':

    chdir('/usr');
    $path = realpath('../sys/kern');

or in anywhere ...

    $path = realpath('/sys/kern');

yields:

    $path eq '/usr/src/sys/kern'

=item abs2rel

    chdir('/usr/local/lib');
    $path = abs2rel('/usr/src/sys');

or in anywhere ...

    $path = abs2rel('/usr/src/sys', '/usr/local/lib');

yields:

    $path eq '../../src/sys'

Similarly,

    $path1 = abs2rel('/usr/src/sys', '/usr');
    $path2 = abs2rel('/usr/src/sys', '/usr/src/sys');

yields:

    $path1 eq 'src/sys'
    $path2 eq '.'

=item rel2abs

    chdir('/usr/local/lib');
    $path = rel2abs('../../src/sys');

or in anywhere ...

    $path = rel2abs('../../src/sys', '/usr/local/lib');

yields:

    $path eq '/usr/src/sys'

Similarly,

    $path = rel2abs('src/sys', '/usr');
    $path = rel2abs('.', '/usr/src/sys');

yields:

    $path eq '/usr/src/sys'

=back

=head1 BUGS

If the base directory includes symbolic links, C<abs2rel> produces the
wrong path.
For example, if '/sys' is a symbolic link to '/usr/src/sys',

    $path = abs2rel('/usr/local/lib', '/sys');

yields:

    $path eq '../usr/local/lib'		# It's wrong!!

You should convert the base directory into a real path in advance.

    $path = abs2rel('/sys/kern', realpath('/sys'));

yields:

    $path eq '../../../sys/kern'	# It's correct but ...

That is correct, but a little redundant. If you wish get the simple
answer 'kern', do the following.

    $path = abs2rel(realpath('/sys/kern'), realpath('/sys'));

C<realpath> assures correct result, but don't forget that C<realpath>
requires that all but the last component of the path exist.

=head1 AUTHOR

Shigio Yamaguchi <shigio@tamacom.com>

=cut
